src/profile-logic/import/dhat.js

```diff
@@ -0,0 +1,324 @@
+  // - bklt=true: a mandatory integer.
+  // - bklt=false: omitted.
+  tg: InstructionCounts,
+
+  // The program points. A mandatory array.
+  pps: ProgramPoint[],
+
+  // Frame table. A mandatory array of strings.
+  // e.g.
+  //  [
+  //   '[root]',
+  //   '0x4AA1D9F: _nl_normalize_codeset (l10nflist.c:332)',
+  //   '0x4A9B414: _nl_load_locale_from_archive (loadarchive.c:173)',
+  //   '0x4A9A2BE: _nl_find_locale (findlocale.c:153)'
+  // ],
+  ftbl: string[],
+|};
+
+type ProgramPoint = {|
+  // Total bytes and blocks. Mandatory integers.
+  tb: Bytes,
+  tbk: Blocks,
+
+  // Total lifetimes of all blocks allocated at this PP.
+  // - bklt=true: a mandatory integer.
+  // - bklt=false: omitted.
+  tl: InstructionCounts,
+
+  // The maximum bytes and blocks for this PP.
+  // - bklt=true: mandatory integers.
+  // - bklt=false: omitted.
+  mb: Bytes,
+  mbk: Blocks,
+
+  // The bytes and blocks at t-gmax for this PP.
+  // - bklt=true: mandatory integers.
+  // - bklt=false: omitted.
+  gb: Bytes,
+  gbk: Blocks,
+
+  // The bytes and blocks at t-end for this PP.
+  // - bklt=true: mandatory integers.
+  // - bklt=false: omitted.
+  eb: Bytes,
+  ebk: Blocks,
+
+  // The reads and writes of blocks for this PP.
+  // - bkacc=true: mandatory integers.
+  // - bkacc=false: omitted.
+  rb: ReadCount,
+  wb: WriteCount,
+
+  // The exact accesses of blocks for this PP. Only used when all
+  // allocations are the same size and sufficiently small. A negative
+  // element indicates run-length encoding of the following integer.
+  // E.g. `-3, 4` means "three 4s in a row".
+  // - bkacc=true: an optional array of integers.
+  // - bkacc=false: omitted.
+  //
+  // e.g. [5, -3, 4, 2]
+  acc: number[],
+
+  // Frames. Each element is an index into the "ftbl" array below.
+  // - All modes: A mandatory array of integers.
+  fs: IndexIntoDhatFrames[],
+|};
+
+// All units of time are in instruction counts.
+// Per: https://valgrind.org/docs/manual/dh-manual.html
+//   As with the Massif heap profiler, DHAT measures program progress by counting
+//   instructions, and so presents all age/time related figures as instruction counts.
+//   This sounds a little odd at first, but it makes runs repeatable in a way which
+//   is not possible if CPU time is used.
+type InstructionCounts = number;
+type Blocks = number;
+type IndexIntoDhatFrames = number;
+type ReadCount = number;
+type WriteCount = number;
+
+export function attemptToConvertDhat(json: mixed): Profile | null {
+  if (!json || typeof json !== 'object') {
+    return null;
+  }
+
+  const { dhatFileVersion } = json;
+  if (typeof dhatFileVersion !== 'number') {
+    // This is not a dhat file.
+    return null;
+  }
+
+  if (dhatFileVersion !== 2) {
+    throw new Error(
+      `This importer only supports dhat version 2. The file provided was version ${dhatFileVersion}.`
+    );
+  }
+  const dhat = coerce<mixed, DhatJson>(json);
+
+  const profile = getEmptyProfile();
```

Comment:
I would have to rewrite it to target a Gecko Profile version, which doesn't have the allocation tables, so I'd prefer targeting the processed version. Otherwise we could start copying out the types for specific profile versions. I'd prefer to have this type checked.

I know Markus prefers targeting old Gecko versions, but I somewhat prefer targeting the current processed version. I know this adds to the maintenance burden for the importers though.

If there is some breaking change that creates a burden, it wouldn't be unreasonable to split this off to the last version where it was working. The following functions should be inlined in that case, and the code should be similar to the upgraders, in that you don't modify the data structures after that.

 * getEmptyProfile,
 * getEmptyThread
 * getEmptyUnbalancedNativeAllocationsTable

Reply:
I'm fine with this approach; can you please make it explicit in a comment at the top of the file?

Reply:
Done.